src/main/java/ch/njol/skript/expressions/ExprSignText.java

```diff
@@ -18,159 +18,297 @@
+		switch (mode) {
+			case ADD:
+				List<String> list = Lists.newArrayList(GetSignLines.getLines(GET_LINES));
+				if (multipleLines) {
+					int last = -1; // Last white space character.
+					for (int i = 3; i >= 0; i--) {
+						if (list.get(i).trim().isEmpty()) {
+							last = i;
+						} else {
+							break;
+						}
+					}
+					if (last < 0)
+						return;
+					int index = 0;
+					for (int i = last; i < 4; i++) {
+						if (index > strings.length - 1)
+							continue;
+						list.set(last, strings[index++]);
+					}
+
+					strings = list.toArray(new String[0]);
+					change(ChangeMode.SET, GET_LINES, GET_LINE, SET_LINE, ADVENTURE_SET_LINE, line, strings);
+				} else {
+					change(ChangeMode.SET, GET_LINES, GET_LINE, SET_LINE, ADVENTURE_SET_LINE, line, GetSignLine.getLine(GET_LINE, line) + StringUtils.join(strings));
+				}
+				break;
+			case RESET:
+			case DELETE:
+				change(ChangeMode.SET, GET_LINES, GET_LINE, SET_LINE, ADVENTURE_SET_LINE, line, CollectionUtils.array("", "", "", ""));
+				break;
+			case REMOVE:
+			case REMOVE_ALL:
+				strings = ChangerUtils.handleStringRemove(StringUtils.join(strings, "\n"), strings[0], mode == ChangeMode.REMOVE_ALL).split("\n");
+				//$FALL-THROUGH$
+			case SET:
+				if (multipleLines) {
+					for (int i = 0; i < 4; i++) {
+						String value = strings.length > i ? (String) strings[i] : "";
+						if (serializer != null && ADVENTURE_SET_LINE != null) {
+							AdventureSetSignLine.line(ADVENTURE_SET_LINE, i, serializer.deserialize(BungeeConverter.convert(ChatMessages.parseToArray(value))));
+							continue;
+						}
+						SetSignLine.setLine(SET_LINE, i, value);
+					}
 					break;
-			}
-			if (hasUpdateBooleanBoolean) {
-				try {
-					s.update(false, false);
-				} catch (final NoSuchMethodError err) {
-					hasUpdateBooleanBoolean = false;
-					s.update();
 				}
-			} else {
-				s.update();
-			}
+				SetSignLine.setLine(SET_LINE, line, (String) strings[0]);
+				break;
+			default:
+				break;
 		}
 	}
-	
+
 	@Override
-	public boolean setTime(final int time) {
-		return super.setTime(time, SignChangeEvent.class, block);
+	public boolean setTime(int time) {
+		return super.setTime(time, SignChangeEvent.class, blocks);
 	}
-	
+
+	@Override
+	public boolean isSingle() {
+		return !multipleLines;
+	}
+
+	@Override
+	public Class<? extends String> getReturnType() {
+		return String.class;
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		if (line == null)
+			return "lines of " + blocks.toString(event, debug);
+		return "line " + line.toString(event, debug) + " of " + blocks.toString(event, debug);
+	}
+
+	@FunctionalInterface
+	private interface SetSignLine<T, S> {
+
+		void setLine(int line, @NotNull String value);
+
+		static void setLine(SetSignLine<Integer, String> setLineMethod, int line, String value) {
```

Comment:
What is the point of these static methods? Why not just call `setLineMethod.setLine(line, value)` directly? Rather than all of these interfaces, I think it would be better just to have a Spigot and Adventure implementations of an interface with set/get methods. The Adventure implementation can handle the string->component conversions.

Reply:
It's required to be static, because that's how FunctionalInterface works. One method must be static.

Adventure doesn't exist on Spigot, so you can't rely on Adventure to handle the string to component conversions as the method doesn't exist.

Reply:
There should be no problem to call the interface methods directly? There's no requirement for there be a static method.

If the Adventure implementation is used only on Paper, I don't see why it would be an issue to handle the component parsing in the interface implantation rather than the syntax's change method. I think the implementation right now is somewhat over complicated for what we need

Reply:
I don't agree with it being more complex. It's making it simpler to read. You have to deal with all the provided methods I posted https://github.com/SkriptLang/Skript/pull/5797#issuecomment-1944531763

Reply:
I think it's unnecessary for the Adventure interface to be something we're specifically checking for in change. Two implementations for set line can exist (the adventure one can handle component parsing), and then there is no need for the change method to worry about those details. Additionally, I'm still not seeing why these static methods are necessary. If you have the interface object, why can't you just call the method it has directly?

Reply:
Getting back to this, I understand the different interfaces, and I see their usefulness, but I don't understand why we can't just use the interface directly. That is, instead of:
```java
SetSignLine.setLine(SET_LINE, i, value)
```
just call
```java
SET_LINE.setLine(i, value)
```
Is there a reason this wouldn't work? If there is some sort of error, can you reply with it?